The `IRTCConnectionObserver` class, used to listen to the connection status of the SDK.

### `on_connected`

Occurs when the connection state of the SDK changes to `3`.

```python
def on_connected(self, agora_rtc_conn, conn_info, reason)
```

#### Parameters

|Parameter|Description|
|---------|-----------|
|`agora_rtc_conn`| The RTC connection object. See [`RtcConnection`](#rtcconnection).|
|`conn_info`|Connection information. See [`RTCConnInfo`](#rtcconninfo).|
|`reason`|Connection state change reason:<ul><li>`0`: Establishing a network connection.</li><li>`1`: Successfully joined the channel.</li><li>`2`: Network connection interrupted.</li><li>`3`: Network connection forbidden by the server.</li><li>`4`: Failed to join the channel.</li><li>`5`: Left the channel.</li><li>`6`: Invalid App ID. Please change to a valid App ID and try again.</li><li>`7`: Invalid channel name. Please change to a valid channel name and try again.</li><li>`8`: The generated token is invalid. Generally, the following reasons:<ul><li>App Certificate is enabled on the console, but the Token is not used when joining the channel. When App Certificate is enabled, you must use a Token.</li><li>The uid specified in [`connect`](#connect) when joining the channel does not match the uid used when generating the Token.</li></ul></li><li>`9`: The generated Token is invalid.</li><li></li><li>`10`: This user is banned by the server.</li><li>`11`: SDK attempts to reconnect due to setting a proxy server.</li><li>`12`: The network connection state changes due to updating the Token.</li><li>`13`: The client IP address changes, possibly due to changes in network type or the IP or port of the network operator.</li><li>`14`: SDK connection to the server times out, enters the automatic reconnection state.</li><li>`15`: Rejoined the channel successfully.</li><li>`16`: SDK and the server lose connection.</li><li>`17`: The connection state changes due to the echo test.</li><li>`18`: The user changed the local IP address.</li></ul>|

### `on_disconnected`

Occurs when the connection state of the SDK changes to `1`.

```python
def on_disconnected(self, agora_rtc_conn, conn_info, reason)
```

#### Parameters

|Parameter|Description|
|---------|-----------|
|`agora_rtc_conn`|The RTC connection object. See [`RtcConnection`](#rtcconnection).|
|`conn_info`|The connection information. See [`RTCConnInfo`](#rtcconninfo).|
|`reason`|Connection state change reason:<ul><li>`0`: Establishing a network connection.</li><li>`1`: Successfully joined the channel.</li><li>`2`: Network connection interrupted.</li><li>`3`: Network connection forbidden by the server.</li><li>`4`: Failed to join the channel.</li><li>`5`: Left the channel.</li><li>`6`: Invalid App ID. Please change to a valid App ID and try again.</li><li>`7`: Invalid channel name. Please change to a valid channel name and try again.</li><li>`8`: The generated token is invalid. Generally, the following reasons:<ul><li>App Certificate is enabled on the console, but the Token is not used when joining the channel. When App Certificate is enabled, you must use a Token.</li><li>The uid specified in [`connect`](#connect) when joining the channel does not match the uid used when generating the Token.</li></ul></li><li>`9`: The generated Token is invalid.</li><li></li><li>`10`: This user is banned by the server.</li><li>`11`: SDK attempts to reconnect due to setting a proxy server.</li><li>`12`: The network connection state changes due to updating the Token.</li><li>`13`: The client IP address changes, possibly due to changes in network type or the IP or port of the network operator.</li><li>`14`: SDK connection to the server times out, enters the automatic reconnection state.</li><li>`15`: Rejoined the channel successfully.</li><li>`16`: SDK and the server lose connection.</li><li>`17`: The connection state changes due to the echo test.</li><li>`18`: The user changed the local IP address.</li></ul>|

### `on_token_privilege_will_expire`

Occurs when the token will expire in 30 seconds.

```python
def on_token_privilege_will_expire(self, agora_rtc_conn, token)
```

In the call process, if the Token is about to expire, the SDK triggers this callback 30 seconds before the Token expires to remind the app to update the Token.

When you receive this callback, you need to generate a new Token on the server side and call [`renew_token`](#renew_token) to pass the newly generated Token to the SDK.

#### Parameters

| Parameter | Description |
|---------------|------------|
| `agora_rtc_conn` | RTC connection object. See [`RtcConnection`](#rtcconnection). |
| `token` | The token that is about to expire. |

### `on_token_privilege_did_expire`

Occurs when the token expires.

```python
def on_token_privilege_did_expire(self, agora_rtc_conn)
```

In the call process, if the token expires, the SDK triggers this callback to remind the app to update the Token.

When you receive this callback, you need to generate a new Token on the server side and call [`connect`](#connect) to reconnect the channel.

#### Parameters

| Parameter | Description |
|---------------|------------|
| `agora_rtc_conn` | RTC connection object. See [`RtcConnection`](#rtcconnection). |

### `on_user_joined`

Occurs when a remote user joins the channel. You can get the remote user's ID through this callback.

```python
def on_user_joined(self, agora_rtc_conn, user_id)
```

#### Parameters

| Parameter | Description |
|---------------|------------|
| `agora_rtc_conn` | RTC connection object. See [`RtcConnection`](#rtcconnection). |
| `user_id` | The remote user's ID. |


### `on_user_left`

Occurs when a remote user leaves the channel. You can get the reason for the remote user leaving the channel through the `reason` parameter.

```python
def on_user_left(self, agora_rtc_conn, user_id, reason)
```

#### Parameters

|Parameter|Description|
|---------|-----------|
|`agora_rtc_conn`|The RTC connection object. See [`RtcConnection`](#rtcconnection).|
|`user_id`|The remote user's ID.|
|`reason`|Connection state change reason:<ul><li>`0`: Establishing a network connection.</li><li>`1`: Successfully joined the channel.</li><li>`2`: Network connection interrupted.</li><li>`3`: Network connection forbidden by the server.</li><li>`4`: Failed to join the channel.</li><li>`5`: Left the channel.</li><li>`6`: Invalid App ID. Please change to a valid App ID and try again.</li><li>`7`: Invalid channel name. Please change to a valid channel name and try again.</li><li>`8`: The generated token is invalid. Generally, the following reasons:<ul><li>App Certificate is enabled on the console, but the Token is not used when joining the channel. When App Certificate is enabled, you must use a Token.</li><li>The uid specified in [`connect`](#connect) when joining the channel does not match the uid used when generating the Token.</li></ul></li><li>`9`: The generated Token is invalid.</li><li></li><li>`10`: This user is banned by the server.</li><li>`11`: SDK attempts to reconnect due to setting a proxy server.</li><li>`12`: The network connection state changes due to updating the Token.</li><li>`13`: The client IP address changes, possibly due to changes in network type or the IP or port of the network operator.</li><li>`14`: SDK connection to the server times out, enters the automatic reconnection state.</li><li>`15`: Rejoined the channel successfully.</li><li>`16`: SDK and the server lose connection.</li><li>`17`: The connection state changes due to the echo test.</li><li>`18`: The user changed the local IP address.</li></ul>|

### `on_stream_message_error`

Occurs when a data stream message error occurs. This callback indicates that the local user has not received the stream message sent by the remote user using the [`send_stream_message`](#send_stream_message) method.

```python
def on_stream_message_error(self, user_id: str, stream_id: int, code: int, missed: int, cached: int)
```

#### Parameters

| Parameter        | Description  |
|-------------|----------|
| `user_id`   | The user ID that sends the message.              |
| `stream_id` | The stream ID of the received message.         |
| `code`      | The code of the error that has occurred.                |
| `missed`    | The number of lost messages.                     |
| `cached`    | The number of messages cached when the data stream is interrupted.          |


### `on_error`

Occurs when there is an error. This callback indicates that an error related to the network or media occurred during the SDK runtime. Generally, the error reported by the SDK means that the SDK cannot automatically recover and requires intervention from the App or user prompt.

```python
def on_error(self, agora_rtc_conn, error_code, error_msg)
```

#### Parameters

| Parameter        | Description  |
|-------------|----------|
| `agora_rtc_conn`   | The RTC connection object. See [`RtcConnection`](#rtcconnection).              |
| `error_code` | The error code.         |
| `error_msg`      | The error description.                |